### PR TITLE
components: handle invalid utf8 characters in `produce_str_with_specified_tp` func

### LINKED ISSUE #12542: TiKV panic when `cast` func pushdown with process invalid utf8 character
### What version of TiKV are you using?

v6.0.0

### Steps to reproduce

```
set @@tidb_skip_utf8_check = 1
create table t (a char(10) charset latin1, b int);
insert into t values (0x90, 1);
select a from t where cast(a as char(10) charset utf8) = a limit 1;
```

### What did you expect?

not panic

### What did happened?

panic

```
[2022/05/17 10:22:10.919 +08:00] [FATAL] [lib.rs:468] ["called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 0, error_len: Some(1) }"] [backtrace="   0: backtrace::capture::Backtrace::new\n   1: tikv_util::set_panic_hook::{{closure}}\n   2: std::panicking::rust_panic_with_hook\n   3: std::panicking::begin_panic_handler::{{closure}}\n   4: std::sys_common::backtrace::__rust_end_short_backtrace\n   5: _rust_begin_unwind\n   6: core::panicking::panic_fmt\n   7: core::result::unwrap_failed\n   8: tidb_query_expr::impl_cast::cast_as_string_helper\n   9: tidb_query_expr::impl_cast::cast_string_as_string_fn_meta::run\n  10: tidb_query_expr::types::expr_eval::<impl tidb_query_expr::types::expr::RpnExpression>::eval_decoded\n  11: <tidb_query_common::execute_stats::WithSummaryCollector<C,T> as tidb_query_executors::interface::BatchExecutor>::next_batch\n  12: <tidb_query_common::execute_stats::WithSummaryCollector<C,T> as tidb_query_executors::interface::BatchExecutor>::next_batch\n  13: tidb_query_executors::runner::BatchExecutorsRunner<SS>::internal_handle_request\n  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n  15: <tikv::coprocessor::interceptors::tracker::Tracker<F> as core::future::future::Future>::poll\n  16: <tikv::coprocessor::interceptors::concurrency_limiter::ConcurrencyLimiter<PF,F> as core::future::future::Future>::poll\n  17: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n  18: yatp::task::future::RawTask<F>::poll\n  19: <yatp::task::future::Runner as yatp::pool::runner::Runner>::handle\n  20: std::sys_common::backtrace::__rust_begin_short_backtrace\n  21: core::ops::function::FnOnce::call_once{{vtable.shim}}\n  22: std::sys::unix::thread::Thread::new::thread_start\n  23: __pthread_deallocate\n"] [location=components/tidb_query_datatype/src/codec/convert.rs:716] [thread_name=unified-read-pool-5]
```

### PR DESCRIPTION
**Pull request by Defined2014**
### What is changed and how it works?

Issue Number: Close #12542

What's Changed:

```commit-message
handle invalid utf8 characters in `produce_str_with_specified_tp` func
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
fix panic when `cast` function pushdown to tikv with invalid utf8 characters.
```

### CODE REVIEW

**Comment by Defined2014** [comment]
/cc @sticnarf @5kbpers 
PTAL

File: components/tidb_query_datatype/src/codec/convert.rs [hunk 1]
```diff
@@ -713,19 +709,23 @@ pub fn produce_str_with_specified_tp<'a>(
...
-            let mut indices = s.char_indices().skip(flen);
-            indices.next().map(|(truncate_pos, _)| {
-                let char_count = flen + 1 + indices.count();
-                (char_count, truncate_pos)
-            })
+            let s = &String::from_utf8_lossy(&s);
+            let mut truncate_pos = 0 as usize;
+            let mut char_indices = s.char_indices();
+            for _ in 0..flen {
+                let utf8_char = char_indices.next();
```
Comment (5kbpers) [inline components/tidb_query_datatype/src/codec/convert.rs:716]: could replace this `for loop` with `char_indices.take(flen).map(...)`

Comment (Defined2014) [reply components/tidb_query_datatype/src/codec/convert.rs]: Done

File: components/tidb_query_datatype/src/codec/convert.rs [hunk 2]
```diff
@@ -713,19 +709,21 @@ pub fn produce_str_with_specified_tp<'a>(
...
-            // so we can unwrap directly here because we need the `s` represent a valid str
-            let s: &str = std::str::from_utf8(s.as_ref()).unwrap();
-            let mut indices = s.char_indices().skip(flen);
-            indices.next().map(|(truncate_pos, _)| {
-                let char_count = flen + 1 + indices.count();
-                (char_count, truncate_pos)
-            })
+            let s = &String::from_utf8_lossy(&s);
+            let mut truncate_pos = 0;
+            s.char_indices().take(flen).for_each(|(_, utf8_char)| {
```
Comment (sticnarf) [inline components/tidb_query_datatype/src/codec/convert.rs:714]:
```suggestion
            s.chars().take(flen).for_each(|utf8_char| {
```

File: components/tidb_query_datatype/src/codec/convert.rs [hunk 3]
```diff
@@ -713,19 +709,21 @@ pub fn produce_str_with_specified_tp<'a>(
...
+                if utf8_char == char::REPLACEMENT_CHARACTER {
+                    truncate_pos += 1;
+                } else {
+                    truncate_pos += utf8_char.len_utf8();
+                }
+            });
+            (s.char_indices().count(), truncate_pos)
```
Comment (sticnarf) [inline components/tidb_query_datatype/src/codec/convert.rs:721]:
```suggestion
            (s.chars().count(), truncate_pos)
```

File: components/tidb_query_datatype/src/codec/convert.rs [hunk 4]
```diff
@@ -713,19 +709,21 @@ pub fn produce_str_with_specified_tp<'a>(
...
         };
-        if truncate_info.is_none() {
+        let (char_count, truncate_pos) = truncate_info;
```
Comment (sticnarf) [inline components/tidb_query_datatype/src/codec/convert.rs:723]: You can just move this to L711:

```
let (char_count, truncate_pos) = {
...
}
```

**Review by sticnarf — APPROVED** [review]
LGTM

**Review by 5kbpers — APPROVED** [review]

**Pull request merged**